src/thunderbots/software/ai/hl/stp/evaluation/enemy_threat.cpp

```diff
@@ -0,0 +1,169 @@
+    //
+    // This algorithm essentially treats the team of robots like a Directed Acyclic
+    // Graph and "traverses" the graph to find the shortest path to the robot
+
+    // The robots that could have the ball and make a pass at each iteration. They
+    // are on the "frontier" of the graph search
+    std::vector<Robot> current_passers{initial_passer};
+    // The remaining robots we haven't checked yet
+    std::vector<Robot> unvisited_robots = team.getAllRobots();
+    // Remove the robot already in the current_passers since it starts off as visited
+    unvisited_robots.erase(
+        std::remove(unvisited_robots.begin(), unvisited_robots.end(), initial_passer),
+        unvisited_robots.end());
+
+    // On each iteration, check what robots can be passed to. These receivers will
+    // become the passers on the next iteration. This is like expanding the frontier
+    // of the graph
+    //
+    // We continue to iterate and check greater numbers of passes until one of the
+    // following cases:
+    // * There are no more passers at the end of the iteration. This means no more
+    //   unvisited robots can be passed to
+    // * There are no more unvisited robots
+    // * We have iterated up to the size of the team. This is a fallback case to
+    //   prevent any infinite loops, just in case
+    //
+    // The pass_num starts at one since the 0-th pass would be when the robot already
+    // has the ball
+    for (int pass_num = 1; pass_num < team.numRobots() && !current_passers.empty() &&
+                           !unvisited_robots.empty();
+         pass_num++)
+    {
+        // A comparator for the map below that stores passer and receiver robots.
+        // This comparator is necessary for the Robot class to be used as a key in the
+        // map. This comparator compares robots by ID.
+        // See
+        // https://stackoverflow.com/questions/6573225/what-requirements-must-stdmap-key-classes-meet-to-be-valid-keys
+        // and
+        // https://stackoverflow.com/questions/5733254/how-can-i-create-my-own-comparator-for-a-map
+        //
+        // This "custom" comparator is defined here rather than in the Robot class
+        // as the '<' operator because there are many possible ways to order robots,
+        // so it doesn't really make sense to define a single "correct" way in the
+        // class, so we just define it as we need it here.
+        struct cmpRobotByID
+        {
+            bool operator()(const Robot& r1, const Robot& r2) const
+            {
+                return r1.id() < r2.id();
+            }
+        };
+
+        // Store a map of robots that can receive the ball, and the list of all robots
+        // that could pass to them
+        std::map<Robot, std::vector<Robot>, cmpRobotByID> receiver_passer_pairs;
+
+        // For each of the current passers, check which unvisited robots they can
+        // pass to
+        for (const auto& current_passer : current_passers)
```

Comment:
Could we make this `for` loop into a function like `std::map<Robot, std::vector<Robot>, cmpRobotByID> findAllRecieverPasserPairs(std::vector<Robot> robots)`? Would probably want to move some of doc comments along with it.

Reply:
I separated this loop into its own function and added more unit tests for it.